### PR TITLE
Integrate macOS tokens for Button

### PR DESCRIPTION
**Pull request by chiuam**
### Platforms Impacted
- [ ] iOS
- [X] macOS
- [ ] win32 (Office)
- [ ] windows
- [ ] android

### Description of changes

Mac tokens were added in #1130, let's consume them in Button.  I also updated the button token values to use Fluent Apple palette colors. 

Updating these token values:
| Token (light-macos) | Before                                       | After                                      |
|---|--------------------------------|---------------------------|
|neutralBackgroundDisabled | #f0f0f0 | #E0E0E0 |
|brandBackground.fillColorRest | #185abd | #0078D4 |
|brandBackground.fillColorHover | #1651aa | #106EBE |
|brandBackground.fillColorPressed | #0e336a | # 004578 |
|brandBackground.fillColorSelected | #13458f | #005A9E |

| Token (dark-macos) | Before                                       | After                                      |
|---|--------------------------------|---------------------------|
|neutralBackgroundDisabled | #141414 | #D6D6D6 |
|neutralStrokeDisabled | #424242 | #525252 |
|brandBackground.fillColorRest | #185abd | #0086F0 |
|brandBackground.fillColorHover | #1651aa | #0078D4 |
|brandBackground.fillColorPressed | #0e336a | #6CB8F6 |
|brandBackground.fillColorSelected | #13458f | #3AA0F3 |

### Verification

| Button | Before                                       | After                                      |
|---|--------------------------------|---------------------------|
| Light |image| image|
| Dark |image| image|
| High contrast light |image|image|
| High contrast dark | image| image|

| Experimental Button | Before                                       | After                                      |
|---|--------------------------------|---------------------------|
| Light | image| image|
| Dark | image | image|
| High contrast light | image | image|
| High contrast dark |image | image|

Primary button hover/press state

https://user-images.githubusercontent.com/67026167/146228956-8e844117-a6cc-41d6-8231-3197cc02f73b.mov

### Pull request checklist

This PR has considered (when applicable):
- [ ] Automated Tests
- [ ] Documentation and examples
- [ ] Keyboard Accessibility
- [ ] Voiceover
- [ ] Internationalization and Right-to-left Layouts

### CODE REVIEW

**Comment by harrieshin** [comment]
screen shot of dark mode, increase contrast mode?

**Comment by harrieshin** [comment]
might be worth actually having before and after table to make sure things didn't change as intented

File: packages/theming/theme-tokens/src/dark-macos/reactnative/tokens-aliases.json [hunk 1]
```diff
@@ -1,262 +1,262 @@
...
+    "fillColorSelected": "#383838"
+  },
+  "neutralBackground2": {
+    "fillColorRest": "#1f1f1f",
+    "fillColorHover": "#333333",
+    "fillColorPressed": "#141414",
+    "fillColorSelected": "#2e2e2e"
+  },
+  "neutralBackground3": {
+    "fillColorRest": "#555555",
```
Comment (harrieshin) [inline packages/theming/theme-tokens/src/dark-macos/reactnative/tokens-aliases.json:107]: explicitly write them in your PR description for history

File: packages/theming/theme-tokens/src/dark-macos/reactnative/tokens-aliases.json [hunk 2]
```diff
@@ -1,262 +1,262 @@
...
+    "fillColorRest": "#d6d6d6",
+    "fillColorHover": "#ffffff",
+    "fillColorPressed": "#ffffff",
+    "fillColorBrandHover": "#d2e0f4",
+    "fillColorBrandPressed": "#2e6ac5",
+    "fillColorBrandSelected": "#d2e0f4",
+    "fillColorSelected": "#ffffff"
+  },
+  "neutralForeground3": {
+    "fillColorRest": "#FFFFFF",
```
Comment (harrieshin) [inline packages/theming/theme-tokens/src/dark-macos/reactnative/tokens-aliases.json:149]: explicitly all out this change in your pr description for history

File: packages/theming/theme-tokens/src/light-macos/reactnative/tokens-aliases.json [hunk 3]
```diff
@@ -1,262 +1,262 @@
...
+    "fillColorSelected": "#ebebeb"
+  },
+  "neutralBackground2": {
+    "fillColorRest": "#fafafa",
+    "fillColorHover": "#f0f0f0",
+    "fillColorPressed": "#dbdbdb",
+    "fillColorSelected": "#e6e6e6"
+  },
+  "neutralBackground3": {
+    "fillColorRest": "#0000000C",
```
Comment (harrieshin) [inline packages/theming/theme-tokens/src/light-macos/reactnative/tokens-aliases.json:107]: first time seeing this hex but ok

Comment (mischreiber) [reply packages/theming/theme-tokens/src/light-macos/reactnative/tokens-aliases.json:107]: Actually that's a good point: I don't see this value in the global tokens; seems like it should be `#F5F5F5`. Any idea what happened here @chiuam @rurikoaraki?

**Comment by Saadnajmi** [comment]
Since this touches the apple theme, can we verify that the non experimental button is also unaffected?

File: packages/theming/theme-tokens/src/index.ts [hunk 4]
```diff
@@ -1,3 +1,4 @@
 export { default as globalTokens } from './tokens-global';
+export { default as aliasTokens } from './light-macos/tokens-alias';
```
Comment (rurikoaraki) [inline packages/theming/theme-tokens/src/index.ts:2]: I would rather us expose alias tokens through an implementation of `getAliasTokens` per platform. That way we allow for the alias tokens to change depending on light/dark/HC mode, and there's only one point exposed to get the alias tokens. If we export alias tokens directly like this, anyone else who imports `aliasTokens` instead of using `getAliasTokens` would only get the macOS light version ones.

File: packages/theming/theme-tokens/src/light-macos/reactnative/tokens-aliases.json (hunk 3 above)
Comment (rurikoaraki) [reply packages/theming/theme-tokens/src/light-macos/reactnative/tokens-aliases.json:107]: @chiuam any ideas? It doesn't look like the input file was touched, so I'm not sure why the output files were changed

**Review by rurikoaraki — CHANGES_REQUESTED** [review]
Would like to see us use `getAliasTokens` instead of exporting alias tokens directly

File: packages/theming/theme-tokens/src/index.ts (hunk 4 above)
Comment (chiuam) [reply packages/theming/theme-tokens/src/index.ts:2]: @rurikoaraki Thank you for the suggestion, added `getAliasTokens` implementation for macOS in 80f44e276ce34b9db6a78deeae6350eeb5a4e24b. Please take a second look when you get a chance.

Comment (rurikoaraki) [reply packages/theming/theme-tokens/src/index.ts:2]: Thanks @chiuam! Is there a reason to keep this implementation separate from the other `getAliasTokens`? Could we make a `getAliasTokens.macos.ts` so that `getAliasTokens` changes what it does on macos automatically instead of exporting an additional function?

Office has its own version as the function signature has to be different from the base `getAliasTokens`, but it looks like the Mac one can match the signature, unless you think this will change in the future.

File: packages/theming/theme-tokens/src/getMacOSAliasTokens.ts [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+import { AppearanceOptions } from '@fluentui-react-native/theme-types';
+
+export function getMacOSAliasTokens(mode: AppearanceOptions) {
+  if (mode === 'light') {
+    return macOSLightAliasTokens.default;
+  } else if (mode === 'dark') {
+    return macOSDarkAliasTokens.default;
+  } else if (mode === 'highContrast') {
+    return hcAliasTokens.default;
+  }
```
Comment (rurikoaraki) [inline packages/theming/theme-tokens/src/getMacOSAliasTokens.ts:13]: nit: we can `assertNever` here to make sure we get a build failure is a new AppearanceOption is added

File: packages/theming/theme-tokens/src/index.ts (hunk 4 above)
Comment (chiuam) [reply packages/theming/theme-tokens/src/index.ts:2]: Oh yeah, I didn't think of that. Fixed in b0aa4517334d4780e2c24a1236028f4730394a74

Comment (chiuam) [reply packages/theming/theme-tokens/src/index.ts:2]: Removed aliasTokens export in 7492c9ef0313fe93d0017ba534cfbb827efa69e4

File: packages/theming/theme-tokens/src/getMacOSAliasTokens.ts (hunk 5 above)
Comment (chiuam) [reply packages/theming/theme-tokens/src/getMacOSAliasTokens.ts]: Added, I also removed the default case as it is unreachable I think...

File: packages/theming/apple-theme/src/appleColors.macos.ts [hunk 6]
```diff
@@ -259,21 +264,13 @@ function getFluentUIApplePalette(): FluentUIApplePalette {
...
-    neutralStroke2: '#00000026',
-    neutralForeground3: DynamicColorMacOS({
-      light: '#272727',
-      dark: '#FFFFFF',
-    }),
-    neutralBackground3: DynamicColorMacOS({
-      light: '#0000000C',
-      dark: '#555555',
-    }),
+    neutralBackgroundInverted: macOSAliasColorTokens.neutralBackgroundInverted,
```
Comment (chiuam) [inline packages/theming/apple-theme/src/appleColors.macos.ts:267]: It's kind of redundant to redefine the colors here, but let's keep it for now until we have all the colors updated?

File: packages/theming/apple-theme/src/appleColors.macos.ts [hunk 7]
```diff
@@ -332,40 +325,40 @@ export function fallbackApplePalette(): ThemeColorDefinition {
...
     primaryButtonTextDisabled: fluentUIApple.brandForegroundDisabled,
 
     accentButtonBackground: fluentUIApple.communicationBlue,
-    accentButtonText: fluentUIApple.neutralInverted,
+    accentButtonText: macOSAliasColorTokens.neutralForegroundInverted,
 
     menuBackground: 'transparent',
     menuDivider: applePlatform.separatorColor,
-    menuIcon: neutralForeground3, //GH:728 Icon doesn't support PlatformColor
+    menuIcon: macOSAliasColorTokens.neutralForeground3,
```
Comment (chiuam) [inline packages/theming/apple-theme/src/appleColors.macos.ts:361]: @Saadnajmi is it appropriate to not use PlatformColor here and use alias tokens instead?

File: packages/theming/theme-tokens/src/index.ts (hunk 4 above)
Comment (rurikoaraki) [reply packages/theming/theme-tokens/src/index.ts:2]: You're still exporting it here though, I think we can remove this, too

File: packages/theming/apple-theme/src/appleColors.macos.ts (hunk 7 above)
Comment (Saadnajmi) [reply packages/theming/apple-theme/src/appleColors.macos.ts:361]: Eventually the token pipeline should have support for platform color, such that this token _is_ a platform color. This specific token is used in svgs so it can't be platform color just yet though.

Comment (Saadnajmi) [reply packages/theming/apple-theme/src/appleColors.macos.ts:361]: I would keep the comment with the bug number

File: packages/theming/theme-tokens/src/index.ts (hunk 4 above)
Comment (chiuam) [reply packages/theming/theme-tokens/src/index.ts:2]: removed, for real this time :)

File: packages/theming/apple-theme/src/appleColors.macos.ts (hunk 7 above)
Comment (chiuam) [reply packages/theming/apple-theme/src/appleColors.macos.ts]: Makes sense, fixed in d2132188efc7e7c453b4eaccff05d9b3f96dd578

File: packages/theming/theme-tokens/src/getMacOSAliasTokens.ts (hunk 5 above)
Comment (Saadnajmi) [reply packages/theming/theme-tokens/src/getMacOSAliasTokens.ts]: `assertNever` was added to check that we never reach theoretically unreachable cases, so I think we would want to add that.

Comment (Saadnajmi) [reply packages/theming/theme-tokens/src/getMacOSAliasTokens.ts]: jk, it's there. My bad

File: packages/theming/theme-tokens/src/index.ts (hunk 4 above)
Comment (chiuam) [reply packages/theming/theme-tokens/src/index.ts]: ^I removed the wrong one, should be fixed now...

File: packages/theming/theme-tokens/src/light-macos/reactnative/tokens-aliases.json (hunk 3 above)
Comment (chiuam) [reply packages/theming/theme-tokens/src/light-macos/reactnative/tokens-aliases.json:107]: I copied the value from the apple theme because it looked a lot closer to the existing button, here's a visual comparison: 
| #F5F5F5                                       | #0000000C                                      |
|----------|--------|
| image | image |

Comment (chiuam) [reply packages/theming/theme-tokens/src/light-macos/reactnative/tokens-aliases.json:107]: I'm not sure whether or not we should go with `#F5F5F5` for buttonBackground, though I do think the hint of grey looks more natural to me. @Saadnajmi thoughts?

Comment (chiuam) [reply packages/theming/theme-tokens/src/light-macos/reactnative/tokens-aliases.json:107]: I think we should follow the [spec](https://www.figma.com/file/ArWQiU7BAlrDGNmbrGEXbA/%5BSpecs%5D-Fluent-vNext-iOS-Components?node-id=744%3A734) for now to move forward and revisit the button background colors when the mac tokens are more finalized.  I'll revert the changes I made to the token values.

Comment (Saadnajmi) [reply packages/theming/theme-tokens/src/light-macos/reactnative/tokens-aliases.json:107]: I copied the apple theme background button color from Fluent UI Apple. I would vote we open a bug on the design token repo or something to fix the macOS token, and follow spec. Especially if both options look acceptable in production, which it seems to be

File: packages/theming/apple-theme/src/appleColors.macos.ts (hunk 6 above)
Comment (Saadnajmi) [reply packages/theming/apple-theme/src/appleColors.macos.ts:267]: There will probably be lots of redundancy till we're fully integrated. Good to keep track where they are so we know to remove them, especially for the sake of bundle size.

Comment (Saadnajmi) [reply packages/theming/apple-theme/src/appleColors.macos.ts:267]: I think if we do move to fully design token based, then we would get rid of the apple theme altogether, and just have it be part of the "Fluent Theme" or "Default Theme", since we don't have native code for this theme anyway.

File: packages/theming/theme-tokens/src/dark-macos/reactnative/tokens-aliases.json (hunk 1 above)
Comment (rurikoaraki) [inline packages/theming/theme-tokens/src/dark-macos/reactnative/tokens-aliases.json:107]: This difference is still here

Can the changes to these files be fully reverted? Or can we update the input files instead if these entries are supposed to change? This file is auto-generated, we should not be editing it directly (same with the light version of this file)

Comment (rurikoaraki) [reply packages/theming/theme-tokens/src/dark-macos/reactnative/tokens-aliases.json:107]: (same comment for neutralForeground3)

File: packages/theming/theme-tokens/src/index.ts (hunk 4 above)
Comment (rurikoaraki) [reply packages/theming/theme-tokens/src/index.ts]: Thanks!

File: packages/theming/theme-tokens/src/dark-macos/reactnative/tokens-aliases.json (hunk 1 above)
Comment (chiuam) [reply packages/theming/theme-tokens/src/dark-macos/reactnative/tokens-aliases.json:107]: I decided to revert the changes here to follow the spec and do a followup once mac tokens are more finalized. I'll update the PR description to include screenshots of before & after

**Review by Saadnajmi — CHANGES_REQUESTED** [review]
Cloned this locally and the different background color is actually slightly off-putting to my eye. I wouldn't block on that, except it stays light colored in dark mode, which is bad.  I'm not sure anyone is using the default button, but I think it would be best not to cause a regression.

image

File: packages/theming/theme-tokens/src/dark-macos/reactnative/tokens-aliases.json (hunk 1 above)
Comment (chiuam) [reply packages/theming/theme-tokens/src/dark-macos/reactnative/tokens-aliases.json:107]: I did `git checkout origin/master <tokens-aliases.json>`, that should revert the file to its original state.

File: packages/theming/theme-tokens/src/dark-macos/reactnative/tokens-aliases.json [hunk 8]
```diff
@@ -1,262 +1,262 @@
 {
-	"body": {
```
Comment (rurikoaraki) [inline packages/theming/theme-tokens/src/dark-macos/reactnative/tokens-aliases.json:2]: I don't think this file should be changed, either. Could you revert the changes in this file as well?

**Comment by chiuam** [comment]
> Cloned this locally and the different background color is actually slightly off-putting to my eye. I wouldn't block on that, except it stays light colored in dark mode, which is bad. I'm not sure anyone is using the default button, but I think it would be best not to cause a regression.
> 
> image

Good catch @Saadnajmi. I wonder if we can detect appearance changes without using DynamicColorMacOS, looking into this...

File: packages/theming/theme-tokens/src/dark-macos/reactnative/tokens-aliases.json (hunk 8 above)
Comment (chiuam) [reply packages/theming/theme-tokens/src/dark-macos/reactnative/tokens-aliases.json:2]: Oops, forgot to revert this one

**Comment by Saadnajmi** [comment]
> > Cloned this locally and the different background color is actually slightly off-putting to my eye. I wouldn't block on that, except it stays light colored in dark mode, which is bad. I'm not sure anyone is using the default button, but I think it would be best not to cause a regression.
> > image
> 
> Good catch @Saadnajmi. I wonder if we can detect appearance changes without using DynamicColorMacOS, looking into this...

You can use the FURN `getCurrentAppearance` API, which I think the apple theme uses for neutralForeground3

File: packages/theming/theme-tokens/src/dark-macos/reactnative/tokens-aliases.json (hunk 8 above)
Comment (chiuam) [reply packages/theming/theme-tokens/src/dark-macos/reactnative/tokens-aliases.json]: Done, thanks for catching!

**Comment by chiuam** [comment]
> > > Cloned this locally and the different background color is actually slightly off-putting to my eye. I wouldn't block on that, except it stays light colored in dark mode, which is bad. I'm not sure anyone is using the default button, but I think it would be best not to cause a regression.
> > > image
> > 
> > 
> > Good catch @Saadnajmi. I wonder if we can detect appearance changes without using DynamicColorMacOS, looking into this...
> 
> You can use the FURN `getCurrentAppearance` API, which I think the apple theme uses for neutralForeground3

Not sure if I'm missing something, but `getCurrentAppearance` doesn't pick up theme changes without re-bundling

**Review by rurikoaraki — APPROVED** [review]

**Comment by chiuam** [comment]
> screen shot of dark mode, increase contrast mode?

added screenshots to the PR description, please take a second look when possible.

**Comment by chiuam** [comment]
> > > screen shot of dark mode, increase contrast mode?
> > 
> > 
> > added screenshots to the PR description, please take a second look when possible.
> 
> The high contrast light mode looks off... why does the background go away / it have no border ._. Maybe it's because borderColor is set to transparent instead of matching the background? EDIT: It looks like other stuff doesn't work well on increased contrast either, like separator :(

I don't think it's the border color since it was transparent before? Probably it's the high contrast token values are off, though I couldn't find any background colors defined for high contrast in the spec...

Separator uses Button from fluentui/react-native so yeah it's affected as well...

**Comment by harrieshin** [comment]
> > > > > Cloned this locally and the different background color is actually slightly off-putting to my eye. I wouldn't block on that, except it stays light colored in dark mode, which is bad. I'm not sure anyone is using the default button, but I think it would be best not to cause a regression.
> > > > > image
> > > > 
> > > > 
> > > > Good catch @Saadnajmi. I wonder if we can detect appearance changes without using DynamicColorMacOS, looking into this...
> > > 
> > > 
> > > You can use the FURN `getCurrentAppearance` API, which I think the apple theme uses for neutralForeground3
> > 
> > 
> > Not sure if I'm missing something, but `getCurrentAppearance` doesn't pick up theme changes without re-bundling on macOS
> 
> Turns out it was because I had `getCurrentAppearance` as a global variable, so it never got updated when theme changes.

would you please update your PR screenshot with the latest change?

**Comment by harrieshin** [comment]
can you share screenshot on RTL as well please?

**Comment by Saadnajmi** [comment]
> can you share screenshot on RTL as well please?

I think the (main) place RTL support is broken is with icons / SVGs. We don't have a way to say "flip this icon in RTL" yet as we do downstream in office. The two main components this affects is Icon and Shimmer. General layout of JS components should flip properly, as long as they use flex box correctly (i.e:  start/end instead of left/right).

For button, the concern would be setting an icon with directional meaning (like a chevron), and we can work around that by informing clients about our behavior, so they know to adjust properly. I don't expect this PR to break any RTL stuff, for what it's worth.

**Review by Saadnajmi — APPROVED** [review]
Approved with the caveat that we probably need to file a bug for and properly handle High Contrast on macOS separately, as neither our FluentUI Apple Button nor are FluentUI React Native button respond to High Contrast changes :')

**Comment by chiuam** [comment]
> > > > > > Cloned this locally and the different background color is actually slightly off-putting to my eye. I wouldn't block on that, except it stays light colored in dark mode, which is bad. I'm not sure anyone is using the default button, but I think it would be best not to cause a regression.
> > > > > > image
> > > > > 
> > > > > 
> > > > > Good catch @Saadnajmi. I wonder if we can detect appearance changes without using DynamicColorMacOS, looking into this...
> > > > 
> > > > 
> > > > You can use the FURN `getCurrentAppearance` API, which I think the apple theme uses for neutralForeground3
> > > 
> > > 
> > > Not sure if I'm missing something, but `getCurrentAppearance` doesn't pick up theme changes without re-bundling on macOS
> > 
> > 
> > Turns out it was because I had `getCurrentAppearance` as a global variable, so it never got updated when theme changes.
> 
> would you please update your PR screenshot with the latest change?

@harrieshin Sure! Highlighted token color changes & updated screenshots to reflect those changes.
